### PR TITLE
Add support for proactive refreshing

### PR DESCRIPTION
**Pull request by xiangyan99**

### CODE REVIEW

File: sdk/identity/azure-identity/azure/identity/_internal/managed_identity_client.py [hunk 1]
```diff
@@ -64,6 +64,11 @@ def _process_response(self, response: PipelineResponse, request_time: int) -> Ac
 
         expires_on = int(content.get("expires_on") or int(content["expires_in"]) + request_time)
         content["expires_on"] = expires_on
+        if "refresh_in" not in content:
+            refresh_in = expires_on - request_time
+            if refresh_in > 60 * 60 * 2:  # 2 hours
```
Comment (pvaneck) [inline sdk/identity/azure-identity/azure/identity/_internal/managed_identity_client.py:69]: Nit: For the sake of consistency with other languages:
```suggestion
            if refresh_in >= 60 * 60 * 2:  # 2 hours
```

File: sdk/identity/azure-identity/tests/test_managed_identity_client.py [hunk 2]
```diff
@@ -36,15 +36,15 @@ def test_caching():
         request_factory=lambda _, __: HttpRequest("GET", "http://localhost"), transport=transport
     )
 
-    token = client.get_cached_token(scope)
+    token, refresh_on = client.get_cached_token(scope)
     assert not token
```
Comment (pvaneck) [inline sdk/identity/azure-identity/tests/test_managed_identity_client.py:40]: We should probably add some `refresh_on` checking with `assert refresh_on is None` here.

File: sdk/identity/azure-identity/tests/test_managed_identity_client.py [hunk 3]
```diff
@@ -36,15 +36,15 @@ def test_caching():
...
 
     with mock.patch(ManagedIdentityClient.__module__ + ".time.time", lambda: now):
         token = client.request_token(scope)
     assert token.expires_on == expected_expires_on
     assert token.token == expected_token
 
-    token = client.get_cached_token(scope)
+    token, refresh_on = client.get_cached_token(scope)
     assert token.expires_on == expected_expires_on
     assert token.token == expected_token
```
Comment (pvaneck) [inline sdk/identity/azure-identity/tests/test_managed_identity_client.py:49]: Can then also add `assert refresh_on == expected_expires_on` here. Would also be good to check when the expires_in is greater than 2 hours. 

I am thinking we could parametrize this test with something like:

```
@pytest.mark.parametrize("expires_in, expected_refresh_in", ((3600, 3600), (10800, 5400)))
def test_caching(expires_in, expected_refresh_in):
...
...
assert refresh_on == now + expected_refresh_in
```

**Review by pvaneck — COMMENTED** [review]
Feature wise, I think this looks good. I think just some small additions to the tests are needed.

Also, perhaps worth updating the changelog in this PR?

File: sdk/identity/azure-identity/tests/test_managed_identity_client.py (hunk 3 above)
Comment (xiangyan99) [reply sdk/identity/azure-identity/tests/test_managed_identity_client.py:50]: Currently there is no service returns refresh_in property. Let's just do the bare minimal test.

**Review by pvaneck — APPROVED** [review]

**Pull request merged**